File: src/assembler/tokenizer.py
from __future__ import annotations
from typing import List, Union

from shared.registers import register_table
from shared.token import Token, TokenType


def is_name_character(char: str) -> bool:
    return char.isalpha() or char == '_'


def tokenize_operands(operands: str) -> List[Token]:
    tokens: List[Token] = []
    # Remove redundant spaces and add a semicolon at the end in order to make the loop iterate one more time
    operands = operands.strip() + ';'
    if operands == "":
        return tokens
     
    current_token: Union[Token, None] = None
    for char in operands:

        if current_token is not None:

            if current_token.type == TokenType.ADDRESS_GENERIC:
                if char.isdigit():
                    current_token = Token(TokenType.ADDRESS_LITERAL, int(char))
                elif is_name_character(char):
                    current_token = Token(TokenType.ADDRESS_IN_REGISTER, char)
                
                continue   
                
            elif current_token.type == TokenType.ADDRESS_LITERAL:
                if char.isdigit():
                    current_token.value *= 10
                    current_token.value += int(char)
                    continue
                tokens.append(current_token)
                current_token = None 

            elif current_token.type == TokenType.ADDRESS_IN_REGISTER:
                if is_name_character(char):
                    current_token.value += char
                    continue

                if char == ' ':
                    continue
                if char != ']':
                    print(f'Expected a \']\' after address in argument list "{operands}", but \'{char}\' was provided.')
                    exit(1)

                register = register_table.get(current_token.value)
                if register is not None:
                    tokens.append(Token(TokenType.ADDRESS, register))
                else:
                    print(f"Unknown register {current_token.value} in argument list \"{operands}\".")
                    exit(1)
                
                continue

            elif current_token.type == TokenType.NAME:
                if is_name_character(char):
                    current_token.value += char
                    continue
                if char == ':':
                    tokens.append(Token(TokenType.LABEL, current_token.value))
                    current_token = None
                    continue

                register = register_table.get(current_token.value)
                if register is not None:
                    tokens.append(Token(TokenType.REGISTER, register))
                else:
                    tokens.append(current_token)
                current_token = None
        
            elif current_token.type == TokenType.NUMBER:
                if char.isdigit():
                    current_token.value *= 10
                    current_token.value += int(char)
                    continue
                tokens.append(current_token)
                current_token = None


        if char == '[':
            current_token = Token(TokenType.ADDRESS_GENERIC, None)
            continue
        
        if is_name_character(char):
            current_token = Token(TokenType.NAME, char)
            continue

        if char.isdigit():
            current_token = Token(TokenType.NUMBER, int(char))
            continue

        if char == ';':
            break
        if char == ',' or char == ' ':
            continue
        

        # If the character isn't handled, raise an error
        raise ValueError(f"Unhandled character: '{char}' in argument list \"{operands}\".")

    return tokens

File: src/rust/assembler/src/tokenizer.rs
use shared::registers::get_register;
use shared::token::{Token, TokenValue};


fn is_name_character(c: char) -> bool {
    c.is_alphabetic() || c == '_'
}


pub fn tokenize_operands(operands: &String) -> Vec<Token> {

    let tokens: Vec<Token> = Vec::new();

    // Add a semicolon at the end in order to make the loop iterate one more time for simplicity
    operands.push(';');

    let mut current_token: Option<Token> = None;

    for c in operands.chars() {

        if let Some(token) = current_token {

            let token: &mut Token = current_token.as_mut().unwrap();

            match token.value {
                TokenValue::AddressGeneric(value) => {
                    if c.is_digit(10) {
                        current_token = Some(
                            Token::new(TokenValue::AddressLiteral(c.to_digit(10).unwrap() as u64))
                        );
                    }
                    else if is_name_character(c) {
                        current_token = Some(
                            Token::new(TokenValue::AddressInRegister(c))
                        );
                    }

                    continue;
                },
                
                TokenValue::AddressLiteral(value) => {
                    if c.is_digit(10) {
                        *value = *value * 10 + c.to_digit(10).unwrap() as u64;
                        continue;
                    }

                    tokens.push(current_token.take().unwrap());                    
                },

                TokenValue::AddressInRegisterIncomplete(value) => {
                    if is_name_character(c) {
                        value.push(c);
                        continue;
                    }

                    if c == ' ' {
                        continue;
                    }
                    if char != ']' {
                        panic!(format!("Expected ']' after address in argument list \"{}\", but '{}' was provided", operands, c));
                    }

                    if let Some(register) = get_register(value) {
                        tokens.push(Token::new(TokenValue::AddressInRegister(register)));
                        current_token = None;
                    }
                    else {
                        panic!(format!("Unknown register \"{}\" in argument list \"{}\"", value, operands));
                    }

                    continue;
                },

                TokenValue::Name(value) => {
                    if is_name_character(c) {
                        value.push(c);
                        continue;
                    }
                    if c == ':' {
                        tokens.push(Token::new(TokenValue::Label(value)));
                        current_token = None;
                        continue;
                    }

                    if let Some(register) = get_register(value) {
                        tokens.push(Token::new(TokenValue::Register(register)));
                        current_token = None;
                    }
                    else {
                        tokens.push(current_token.take().unwrap());
                    }
                }
                   
                TokenValue::Number(value) => {
                    if c.is_digit(10) {
                        *value = *value * 10 + c.to_digit(10).unwrap() as u64;
                        continue;
                    }

                    tokens.push(current_token.take().unwrap());
                },
                
                _ => { }
                
            }
        }


        if c == '[' {
            current_token = Some(Token::new(TokenType::AddressGeneric(0)));
            continue;
        }

        if is_name_character(c) {
            current_token = Some(Token::new(TokenValue::Name(c.to_string())));
            continue;
        }

        if c.is_digit(10) {
            current_token = Some(Token::new(TokenValue::Number(c.to_digit(10).unwrap() as i64)));
            continue;
        }

        match c {
            ' ' |
            '\t' \
            ' ' 
            => continue,

            ';' => break,

            '[' => {
                current_token = Some(Token::new(TokenType::AddressGeneric(0)));
                continue;
            },

            _ => panic!(format!("Unhaldled character '{}' in operands \"{}\"", c, operands))
        }
    }   

    tokens
}


